### PR TITLE
MSVC build: Add default warning flags to CMake build

### PR DESCRIPTION
**Pull request by studoot**
This pull request adds warning flags '/W4 /WX' to the MSVC build created with CMake from the CMakeLists.txt files in the build. This also includes code to remove the '/W3' warning level flag that CMake has in its MSVC compiler flag defaults.

There is also code to set the default GCC/Clang compiler flags. These defaults have been taken from the configure.ac in the build.

### CODE REVIEW

**Comment by studoot** [comment]
I've not attempted to fix the new errors caused by these flags, as I'm not sure whether it's preferable to just quieten the relevant warnings or modify the code to eliminate them (if @murraycu can give me some guidance on what's preferred, I'll fix things). As far as I can tell, the 67 warnings raised fall into three categories:

1. `warning C4100: 'type': unreferenced formal parameter` raised on template code with `if constexpr` statements where the parameter isn't used in all paths through the code. This feels like MSVC being extra fussy, as the parameter is clearly used in other instances of the templated code.
2. `warning C4127: conditional expression is constant` raised on code in templates that is dependent on static properties of a template parameter (such as a type's size). These can be fixed using `if constexpr` for the most part.
3. `warning C4244: 'return': conversion from 'int' to 'float', possible loss of data` - one instance in `tests/test_retype_return.cc`. This can be silenced for just that test source file

The full [build log](https://github.com/libsigcplusplus/libsigcplusplus/files/3724507/build.log) is attached.

**Comment by studoot** [comment]
I've added possible fixes for the different warnings. If the fix I've used isn't appropriate, I can re-do.

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,6 +30,23 @@ project (sigc++)
...
+    # Remove the CMake default of /W3 because when you add /W4, MSVC will complain
+    # about two warning level flags
+    string(REGEX REPLACE "(^|[ \t])/W[0-9]($|[ \t])" "\\1\\2" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+    string(REGEX REPLACE "(^|[ \t])/W[0-9]($|[ \t])" "\\1\\2" CMAKE_CXX_FLAGS
+    "${CMAKE_CXX_FLAGS}")
+elseif(CXX_COMPILER_ID)
+endif()
+
+# Add compiler warning flags & turn warnings into errors
+add_compile_options(
```
Comment (murraycu) [inline CMakeLists.txt]: Wow. This really does seem to be the official way to do it. That syntax is hard to like.

I quite like that, in the autotools build, using mm-common, it figures out which of the command line options are actually acceptable to the compiler. Anyway, I'm just grumbling - this seems to be the correct cmake way.

File: CMakeLists.txt [hunk 2]
```diff
@@ -30,6 +30,23 @@ project (sigc++)
 
 set( CMAKE_CXX_STANDARD 17 )
 
+# Turn on warnings for MSVC
+if (MSVC)
+    # Remove the CMake default of /W3 because when you add /W4, MSVC will complain
```
Comment (murraycu) [inline CMakeLists.txt]: I think this cmake PR will provide a better way to do this, so please link to it in the comment:
https://gitlab.kitware.com/cmake/cmake/merge_requests/3250

(Feel free to improve this commit and force push the whole PR branch.)

File: sigc++/tuple-utils/tuple_end.h [hunk 3]
```diff
@@ -56,6 +56,7 @@ tuple_end(T&& t) {
   static_assert(len <= size, "The tuple size must be less than or equal to the length.");
 
   if constexpr(len == 0) {
+    static_cast<void>(t);
```
Comment (murraycu) [inline sigc++/tuple-utils/tuple_end.h]: Please add a comment in the code about why this is here. One day we might be able to remove it, if we have that clue.

File: sigc++/tuple-utils/tuple_for_each.h [hunk 4]
```diff
@@ -103,6 +103,8 @@ tuple_for_each(T&& t, T_extras&&... extras)
   {
     detail::tuple_for_each_impl<T_visitor, size, T_extras...>::tuple_for_each(
       std::forward<T>(t), std::forward<T_extras>(extras)...);
+  } else {
+    static_cast<void>(t);
```
Comment (murraycu) [inline sigc++/tuple-utils/tuple_for_each.h]: Again, this needs a comment, please.

File: sigc++/tuple-utils/tuple_transform_each.h [hunk 5]
```diff
@@ -35,6 +35,7 @@ struct tuple_transform_each_impl {
   static decltype(auto)
   tuple_transform_each(T_current&& t, T_original& t_original) {
     if constexpr(size_from_index == 0) {
+      static_cast<void>(t_original);
```
Comment (murraycu) [inline sigc++/tuple-utils/tuple_transform_each.h]: Add an explanatory comment, please.

File: sigc++/visit_each.h [hunk 6]
```diff
@@ -47,6 +47,8 @@ struct limit_trackable_target
     //Only call action_() if T_Type derives from trackable.
     if constexpr(is_base_of_or_same_v<sigc::trackable, T_type>) {
         std::invoke(action_, type);
+    } else {
+      static_cast<void>(type);
```
Comment (murraycu) [inline sigc++/visit_each.h]: Add an explanatory comment, please.

File: CMakeLists.txt (hunk 1 above)
Comment (studoot) [reply CMakeLists.txt]: CMake has something similar to that for *some* command-line flags (e.g. [setting the C++ standard you want to use](https://cmake.org/cmake/help/v3.1/prop_tgt/CXX_STANDARD.html)), but not warnings...

The other syntax that could be used is something like this:

```CMake
if (MSVC)
   add_compile_options(/W4 /W3)
elseif(GNU)
   add_compile_options(-pedantic -Wall -Wextra ...)
elseif(Clang OR AppleClang)
   add_compile_options(-pedantic -Wall -Wextra ...)
endif()
```

That is probably clearer? As you say, those generator expressions aren't nice...

File: CMakeLists.txt (hunk 2 above)
Comment (studoot) [reply CMakeLists.txt]: That PR has been included in CMake 3.15 (the most recent major release), so I'll link to that. In fact - I can just add a version check in that `if()`, so if you'r eusing a new enough CMake, it's not done.

**Comment by murraycu** [comment]
Ideally the code comments would have been amendments to the original commits, but that is not essential. Many thanks for giving us this ability.

**Pull request merged**